server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java

```diff
@@ -0,0 +1,152 @@
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.db.permission.GroupPermissionDto;
+import org.sonar.db.permission.UserPermissionDto;
+import org.sonar.server.component.ComponentFinder;
+import org.sonar.server.permission.index.PermissionIndexer;
+import org.sonar.server.user.UserSession;
+
+import static java.util.Collections.singletonList;
+import static org.sonar.core.permission.ProjectPermissions.PUBLIC_PERMISSIONS;
+import static org.sonar.server.ws.KeyExamples.KEY_PROJECT_EXAMPLE_001;
+import static org.sonar.server.ws.WsUtils.checkRequest;
+import static org.sonarqube.ws.client.project.ProjectsWsParameters.PARAM_PROJECT;
+
+public class ChangeVisibilityAction implements ProjectsWsAction {
+  private static final String ACTION = "change_visibility";
+  private static final String PARAM_VISIBILITY = "visibility";
+  private static final String PUBLIC_VISIBILITY = "public";
+  private static final String PRIVATE_VISIBILITY = "private";
+  private static final Set<String> ALLOWED_QUALIFIERS = ImmutableSet.of(Qualifiers.PROJECT, Qualifiers.VIEW);
+
+  private final DbClient dbClient;
+  private final ComponentFinder componentFinder;
+  private final UserSession userSession;
+  private final PermissionIndexer permissionIndexer;
+
+  public ChangeVisibilityAction(DbClient dbClient, ComponentFinder componentFinder, UserSession userSession,
+    PermissionIndexer permissionIndexer) {
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.userSession = userSession;
+    this.permissionIndexer = permissionIndexer;
+  }
+
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION)
+      .setDescription("Change visibility of a project or a view.<br/>" +
+        "Requires 'Project administer' permission on the specified project or view<br/>")
+      .setSince("6.4")
+      .setPost(true)
+      .setHandler(this);
+
+    action.createParam(PARAM_PROJECT)
+      .setDescription("Project or view key")
+      .setExampleValue(KEY_PROJECT_EXAMPLE_001)
+      .setRequired(true);
+
+    action.createParam(PARAM_VISIBILITY)
+      .setDescription("new visibility of the project or view")
+      .setPossibleValues(PUBLIC_VISIBILITY, PRIVATE_VISIBILITY)
+      .setRequired(true);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn();
+
+    String projectKey = request.mandatoryParam(PARAM_PROJECT);
+    boolean changeToPrivate = PRIVATE_VISIBILITY.equals(request.mandatoryParam(PARAM_VISIBILITY));
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ComponentDto component = componentFinder.getByKey(dbSession, projectKey);
+      checkRequest(isRoot(component), "Component must either be a project or a view");
+      checkRequest(!changeToPrivate || !Qualifiers.VIEW.equals(component.qualifier()), "Views can't be made private");
```

Comment:
If views are always public, they should either not be allowed for this ws at all, or it should at least be part of the api doc.